File: internal/website/src/types/content.ts
import type { User, Pagination } from './user'

// ContentType represents the type of content.
export type ContentType = 'video' | 'gallery' | 'article'

// ContentStatus represents the publication status of content.
export type ContentStatus = 'draft' | 'published'

// ContentCategory represents the category of content (dynamic, stored as slug string).
export type ContentCategory = string

// Category represents a content category managed in the database.
export interface Category {
  id: string
  name: string
  slug: string
  icon: string
  visible: boolean
  sort_order: number
  content_count: number
  created_at: string
  updated_at: string
}

// SortField represents the field to sort by.
export type SortField = 'created_at' | 'like_count'

// AttachmentType represents the type of an attachment.
export type AttachmentType = 'video' | 'image'

// Attachment represents a file (video, image, etc.) attached to a content.
export interface Attachment {
  id: string
  content_id: string
  title: string
  description: string
  url: string
  type: AttachmentType
  sort_order: number
  is_cover: boolean
  file_size: number
  duration: number
  created_at: string
  updated_at: string
}

// CreateAttachmentArgs represents the fields required to create an attachment.
export interface CreateAttachmentArgs {
  title?: string
  description?: string
  url: string
  type: AttachmentType
  sort_order: number
  is_cover?: boolean
  file_size?: number
  duration?: number
}

// Content represents a piece of content in the system.
export interface Content {
  id: string
  author_id: string
  title: string
  summary: string
  body: string
  cover_url: string
  type: ContentType
  status: ContentStatus
  category: ContentCategory
  tags: string[]
  speaker_id: string
  speaker_name: string
  speaker_bio: string
  like_count: number
  comment_count: number
  created_at: string
  updated_at: string
  author?: User
  speaker?: User
  attachments?: Attachment[]
  liked?: boolean
}

// ListContentsArgs represents the query parameters for listing contents.
export interface ListContentsArgs {
  page?: number
  limit?: number
  category?: ContentCategory
  type?: ContentType
  status?: ContentStatus | 'all'
  keyword?: string
  tag?: string
  sort?: SortField
  author_id?: string
  speaker_id?: string
  followed_by_user_id?: string
}

// ListContentsResponse represents the response for listing contents.
export interface ListContentsResponse {
  contents: Content[]
  pagination: Pagination
}

// CreateContentArgs represents the fields required to create content.
export interface CreateContentArgs {
  title: string
  summary?: string
  body?: string
  cover_url?: string
  type: ContentType
  status?: ContentStatus
  category: ContentCategory
  tags?: string[]
  speaker_id?: string
  speaker_name?: string
  speaker_bio?: string
  media_items?: CreateAttachmentArgs[]
}

// UpdateContentArgs represents the fields that can be updated for content.
export interface UpdateContentArgs {
  title?: string
  summary?: string
  body?: string
  cover_url?: string
  type?: ContentType
  status?: ContentStatus
  category?: ContentCategory
  tags?: string[]
  speaker_id?: string
  speaker_name?: string
  speaker_bio?: string
  media_items?: CreateAttachmentArgs[]
}

// LegacyTalk represents the data structure from the old platform.
export interface LegacyTalk {
  id: string
  title: string
  cover: string
  start_at: string
  description: string
  tags: string[]
  speaker: string
  staff: string
  bio: string
  avatar: string
  broadcast: string
  playback: string
  type: string      // "sharing" or "training"
  volume: string    // e.g., "AI 赋能组织分享会"
  created_at: string
  updated_at: string
}

// ImportContentsArgs represents the request body for importing contents.
export interface ImportContentsArgs {
  contents: LegacyTalk[]
}

// ImportResult represents the result of an import operation.
export interface ImportResult {
  total: number
  imported: number
  skipped: number
  errors: string[]
}

// Comment represents a comment on a content item.
export interface Comment {
  id: string
  content_id: string
  user_id: string
  parent_id: string
  reply_to_id: string
  body: string
  like_count: number
  created_at: string
  updated_at: string
  user?: User
  reply_to?: Comment
  replies?: Comment[]
}

// ListCommentsResponse represents the response for listing comments.
export interface ListCommentsResponse {
  comments: Comment[]
  pagination: Pagination
  liked_comment_ids: string[]
}

// CreateCommentArgs represents the fields required to create a comment.
export interface CreateCommentArgs {
  body: string
  parent_id?: string
  reply_to_id?: string
}

// LikeResponse represents the response after toggling a like.
export interface LikeResponse {
  liked: boolean
  like_count: number
}

File: internal/entity/content.go
package entity

import "time"

// ContentType represents the type of content.
type ContentType string

const (
	// ContentTypeVideo indicates video content with a playlist.
	ContentTypeVideo ContentType = "video"
	// ContentTypeGallery indicates image/short-video gallery content.
	ContentTypeGallery ContentType = "gallery"
	// ContentTypeArticle indicates a long-form article.
	ContentTypeArticle ContentType = "article"
)

// ReservedSlugs contains slugs reserved for content type routes that cannot be used as category slugs.
var ReservedSlugs = map[string]bool{
	"videos":    true,
	"galleries": true,
	"articles":  true,
}

// ContentStatus represents the publication status of content.
type ContentStatus string

const (
	// ContentStatusDraft indicates unpublished draft content.
	ContentStatusDraft ContentStatus = "draft"
	// ContentStatusPublished indicates published content visible to all.
	ContentStatusPublished ContentStatus = "published"
)

// Content represents a piece of content in the system.
type Content struct {
	ID            string        `json:"id"            gorm:"column:id;primaryKey;size:36"`
	AuthorID      string        `json:"author_id"     gorm:"column:author_id;size:36;index:idx_contents_author_id"`
	Title         string        `json:"title"         gorm:"column:title"`
	Summary       string        `json:"summary"       gorm:"column:summary"`
	Body          string        `json:"body"          gorm:"column:body;type:text"`
	CoverURL      string        `json:"cover_url"     gorm:"column:cover_url"`
	Type          ContentType   `json:"type"          gorm:"column:type"`
	Status        ContentStatus `json:"status"        gorm:"column:status;size:32;default:draft;index:idx_contents_status"`
	Category      string        `json:"category"      gorm:"column:category;size:64;index:idx_contents_category"`
	Tags          []string      `json:"tags"          gorm:"column:tags;serializer:json"`
	SpeakerID     string        `json:"speaker_id"    gorm:"column:speaker_id;size:36;index:idx_contents_speaker_id"`
	SpeakerName   string        `json:"speaker_name"  gorm:"column:speaker_name"`
	SpeakerBio    string        `json:"speaker_bio"   gorm:"column:speaker_bio"`
	LikeCount     int64         `json:"like_count"     gorm:"column:like_count;default:0"`
	FavoriteCount int64         `json:"favorite_count" gorm:"column:favorite_count;default:0"`
	CommentCount  int64         `json:"comment_count"  gorm:"column:comment_count;default:0"`
	CreatedAt     time.Time     `json:"created_at"`
	UpdatedAt     time.Time     `json:"updated_at"`

	Author      *User        `json:"author,omitempty"      gorm:"foreignKey:AuthorID"`
	Speaker     *User        `json:"speaker,omitempty"     gorm:"foreignKey:SpeakerID"`
	Attachments []Attachment `json:"attachments,omitempty" gorm:"foreignKey:ContentID"`
}

// TableName specifies the database table name for Content.
func (Content) TableName() string {
	return "contents"
}

// SortField represents the field to sort by.
type SortField string

const (
	// SortByCreatedAt sorts by creation time.
	SortByCreatedAt SortField = "created_at"
	// SortByLikeCount sorts by like count.
	SortByLikeCount SortField = "like_count"
)

// ListContentsArgs represents the query parameters for listing contents.
type ListContentsArgs struct {
	Pagination
	Category         string        `json:"category"            form:"category"`
	Type             ContentType   `json:"type"                form:"type"`
	Status           ContentStatus `json:"status"              form:"status"`
	Keyword          string        `json:"keyword"             form:"keyword"`
	Tag              string        `json:"tag"                 form:"tag"`
	Sort             SortField     `json:"sort"                form:"sort"`
	AuthorID         string        `json:"author_id"           form:"author_id"`
	SpeakerID        string        `json:"speaker_id"          form:"speaker_id"`
	FollowedByUserID string        `json:"followed_by_user_id" form:"followed_by_user_id"`
}

// CreateContentArgs represents the fields required to create content.
type CreateContentArgs struct {
	Title       string                 `json:"title"        binding:"required"`
	Summary     string                 `json:"summary"`
	Body        string                 `json:"body"`
	CoverURL    string                 `json:"cover_url"`
	Type        ContentType            `json:"type"         binding:"required"`
	Status      ContentStatus          `json:"status"`
	Category    string                 `json:"category"     binding:"required"`
	Tags        []string               `json:"tags"`
	SpeakerID   string                 `json:"speaker_id"`
	SpeakerName string                 `json:"speaker_name"`
	SpeakerBio  string                 `json:"speaker_bio"`
	Attachments []CreateAttachmentArgs `json:"attachments"`
	// CreatedAt and UpdatedAt can only be set by admin (e.g. for importing legacy content)
	CreatedAt *time.Time `json:"created_at"`
	UpdatedAt *time.Time `json:"updated_at"`
}

// UpdateContentArgs represents the fields that can be updated for content.
type UpdateContentArgs struct {
	Title       *string                `json:"title"`
	Summary     *string                `json:"summary"`
	Body        *string                `json:"body"`
	CoverURL    *string                `json:"cover_url"`
	Type        *ContentType           `json:"type"`
	Status      *ContentStatus         `json:"status"`
	Category    *string                `json:"category"`
	Tags        []string               `json:"tags"`
	SpeakerID   *string                `json:"speaker_id"`
	SpeakerName *string                `json:"speaker_name"`
	SpeakerBio  *string                `json:"speaker_bio"`
	Attachments []CreateAttachmentArgs `json:"attachments"`
	// CreatedAt and UpdatedAt can only be set by admin (e.g. for importing legacy content)
	CreatedAt *time.Time `json:"created_at"`
	UpdatedAt *time.Time `json:"updated_at"`
}
